### DISCUSSION TITLE
Adguardhome (dns only, not admin interface) through cloudflare tunnels (cloudflared) ?

### DISCUSSION THREAD

**Discussion by Waldorf3** [AdguardTeam/AdGuardHome#5386] [Q&A]
Cloudflare tunnels are a kind of reverse VPN from cloudflare to a home network.  It is a more secure alternative to the traditional problems accessinging a home network with DDNS and port forwarding.  

I have a working setup with wildcard DNS+cert tunnelling traffic to a (docker) nginx-proxy-manager ingress from where I manage which services I will allow access from the public internet, and which automatically handles letsencrypt certificates.  By and large this is a close to zero-maintenance, highly secure access method.

My problem now is I'd like to access the adguardhome server on my home network.  I've eventually managed to have it setup and working perfect from within the home network, and now would like to extend this so I can use it when I am on the public network as well - I just can't get it to work.  

I think it is a matter of setting up cloudflare right, so I'm wondering if anyone has got this working and can provide hints to how the tunnel needs to be setup for dns to work through this tunnel.

I've tried setting up a second public hostname of adguard.mydomain.com pointing to the local IP address of the server running adguardhome, using both http, https and tcp, and port 853, but clients on the public net fails to connect to it. 

I think it is a matter of setting up cloudflare right, so I'm wondering if anyone has got this working and can provide hints to how the tunnel needs to be setup for dns to work through this tunnel.

**Comment by Waldorf3** [comment]
No interest, huh?  Am I the only person in the world wanting to use adguardhome from both my home and the global internet?

**Comment by arihid** [comment]
Just discovered that you can do just that!
Try to connect with Intra (Android, idk about iOS, didn't use them), and set the DNS server to yourdomain.com/dns-query,
You'll need to enable HTTPS in your ADH instance.

![gambar](https://github.com/AdguardTeam/AdGuardHome/assets/82217590/185ee079-2e4f-466d-b65f-b94026d6b707)

172.* comes from cloudflare network.

![Screenshot_2023-11-12-17-09-57-350_app intra-edit](https://github.com/AdguardTeam/AdGuardHome/assets/82217590/d452344c-43df-48c8-9750-b34a850e64ab)


![Screenshot_2023-11-12-17-10-40-399_app intra-edit](https://github.com/AdguardTeam/AdGuardHome/assets/82217590/f38a444b-65e9-4ba1-9d91-de3978f32afd)

**Comment by iwannet** [comment]
How did you do this ? I did this but it just doesnt work. 
![afbeelding](https://github.com/AdguardTeam/AdGuardHome/assets/95164143/87eb6a0b-611d-432b-80aa-cf494162235a)

**Reply by iwannet** [reply]
i enabled encryption in adguardhome using the certificate of my domain.com but when visiting the ip of adguardhome it says that the certificate is only valid for domain.com and not for the ip, maybe thats the problem ?

**Reply by iwannet** [reply]
I can access the admin interface of adgh though

**Reply by arihid** [reply]
This solution is limited to DNS-over-HTTPS, if you can access your web interface from your domain.com, you can access your DoH from domain.com/dns-query. Native Android support is DNS-over-QUIC which I myself had a hard time figuring it out.

In your Cloudflare, you can set the encryption level to whichever you want, but if you decide to use Encrypted Edge (or whatever it is that encrypts the connection from your server to cloudflare is called), you'll need to install the edge certificate from cloudflare (you can find it somewhere in the documentation, I forgot the exact location).

**Reply by iwannet** [reply]
ow i made an new certificate because i said that the cf certificate wasnt valid. 

>  you can access your DoH from domain.com/dns-query

Yeah i tried but it doesnt work, it just doesnt load. But the admin page does load

**Reply by iwannet** [reply]
Now i set  my config like this  ![afbeelding](https://private-user-images.githubusercontent.com/95164143/291053858-87eb6a0b-611d-432b-80aa-cf494162235a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDI4MDgwMTQsIm5iZiI6MTcwMjgwNzcxNCwicGF0aCI6Ii85NTE2NDE0My8yOTEwNTM4NTgtODdlYjZhMGItNjExZC00MzJiLTgwYWEtY2Y0OTQxNjIyMzVhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMTclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjE3VDEwMDgzNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTYwMmJkYWFhMmUxNDhlYmI2OWQ3MDA0ZjdjNWMwMzlhMzkzY2UzMTJhM2Y5YWJhZmVmNzhkMTE3NTA2NTMzMDgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.pmLjXOvKVK-eLbnnHxng_8u4bjS5CjVdvTB3wWmiSsw) and it gives  Bad gateway Error code 502 


And when i do the config but http instead of https i can access the admin-page but not the dns

**Reply by arihid** [reply]
Try to disable no-tls-verify.

**Reply by iwannet** [reply]
where ?

**Reply by iwannet** [reply]
it think its because my certificate is only for my domain and not for the ip adress. because i get this error 

> Websites` prove their identity through certificates. LibreWolf does not trust this website because it uses a certificate that is not valid for 192.168.178.105. The certificate is only valid for the following names: domain
Error code: SSL_ERROR_BAD_CERT_DOMAIN

**Reply by arihid** [reply]
Yes, it might be. And no-tls-verify will skip certificate check, allowing you to connect from origin with any certificate installed.
![Screenshot_2023-12-17-19-17-12-388_com android chrome-edit](https://github.com/AdguardTeam/AdGuardHome/assets/82217590/56b9ff7b-6938-49a1-8cb6-c6ae14920f0b)

**Comment by iwannet** [comment]
Bro thanks it finally works. One question though, how can i disable the access to the adminpage and only allow the use of the dns ?

**Reply by arihid** [reply]
I don't know. I don't even know if Adguard had any rule for that. I guess you can do something with page rule in Cloudflare, you should start there.
One thing for sure is use a good user and password combinantion, bla bla bla, you know, the usual account security lecture.

**Reply by Ankitd013** [reply]
@iwannet You have to add “login.html” in Access -> Applications -> Application domain -> path, and In the firewall rule, add your email address or group (if you have any). Then you can access DNS without authentication but Adguard home protected with Cloudflare auth.

**Reply by iwannet** [reply]
Ow, i just enabled the proxy only for the path `https://dns.server/dns-query` and it works fine to.

**Reply by Ankitd013** [reply]
I am stating that if you wish to disable the admin page while still wanting access to the DNS server, you can do so.

**Comment by Gauvino** [comment]
Hello, can someone please explain me how you did it cause the image cannot be rendered and I don't understand how you did everything, seem's like this issue is missing some info to make the DNS-over-TLS possible with cloudflared

**Reply by arihid** [reply]
You just need to enable HTTPS on your ADH instance and set the hostname in Cloudflare Tunnel like you would. Don't forget to either disable tls verify or install the Cloudflare Edge certificate to avoid 502 error.
